### PR TITLE
Fix Events API handling issue in shared channels

### PR DESCRIPTION
**Pull request by seratch**
This pull request fixes a bug where some Events API handling code (e.g., `message` events) does not properly work in shared channels. All Bolt framework have the same issue. For the details of the issue, refer to  https://github.com/slackapi/java-slack-sdk/issues/629

### Category (place an `x` in each of the `[ ]`)

* [x] `slack_bolt.App` and/or its core components
* [x] `slack_bolt.async_app.AsyncApp` and/or its core components
* [ ] Adapters in `slack_bolt.adapter`
* [ ] Document pages under `/docs`
* [ ] Others

## Requirements (place an `x` in each `[ ]`)

Please read the [Contributing guidelines](https://github.com/slackapi/bolt-python/blob/main/.github/contributing.md) and [Code of Conduct](https://slackhq.github.io/code-of-conduct) before creating this issue or pull request. By submitting, you are agreeing to those rules.

* [x] I've read and understood the [Contributing Guidelines](https://github.com/slackapi/bolt-python/blob/main/.github/contributing.md) and have done my best effort to follow them.
* [x] I've read and agree to the [Code of Conduct](https://slackhq.github.io/code-of-conduct).
* [x] I've run `./scripts/install_all_and_run_tests.sh` after making the changes.

### CODE REVIEW

File: slack_bolt/middleware/authorization/async_multi_teams_authorization.py [hunk 1]
```diff
@@ -54,7 +53,9 @@ async def async_process(
                 req.context.set_authorize_result(auth_result)
                 token = auth_result.bot_token or auth_result.user_token
                 req.context["token"] = token
-                req.context["client"] = create_async_web_client(token)
```
Comment (seratch) [inline slack_bolt/middleware/authorization/async_multi_teams_authorization.py:57]: While writing tests for this pull request, I've found an issue where `context.client` does not respect the given `base_url` etc. As it was a blocker for writing the tests in this PR, I've fixed the issue too.

File: slack_bolt/middleware/authorization/multi_teams_authorization.py [hunk 2]
```diff
@@ -55,7 +55,9 @@ def process(
                 req.context.set_authorize_result(auth_result)
                 token = auth_result.bot_token or auth_result.user_token
                 req.context["token"] = token
-                req.context["client"] = create_web_client(token)
```
Comment (seratch) [inline slack_bolt/middleware/authorization/multi_teams_authorization.py:58]: While writing tests for this pull request, I've found an issue where `context.client` does not respect the given `base_url` etc. As it was a blocker for writing the tests in this PR, I've fixed the issue too.

File: slack_bolt/request/internals.py [hunk 3]
```diff
@@ -110,21 +118,21 @@ def extract_channel_id(payload: Dict[str, Any]) -> Optional[str]:
     return None
 
 
-def build_context(context: BoltContext, payload: Dict[str, Any],) -> BoltContext:
-    enterprise_id = extract_enterprise_id(payload)
+def build_context(context: BoltContext, body: Dict[str, Any]) -> BoltContext:
```
Comment (seratch) [inline slack_bolt/request/internals.py:121]: just to make the name in `internals` code clearer (payload != body)

File: tests/scenario_tests/test_events_shared_channels.py [hunk 4]
```diff
@@ -0,0 +1,514 @@
...
+    def build_headers(self, timestamp: str, body: str):
+        return {
+            "content-type": ["application/json"],
+            "x-slack-signature": [self.generate_signature(body, timestamp)],
+            "x-slack-request-timestamp": [timestamp],
+        }
+
+    valid_event_body = {
+        "token": "verification_token",
+        "team_id": "T_INSTALLED",
```
Comment (seratch) [inline tests/scenario_tests/test_events_shared_channels.py:57]: for `app_mention` events, the top-level team_id is the one for an installed workspace.

File: tests/scenario_tests/test_events_shared_channels.py [hunk 5]
```diff
@@ -0,0 +1,514 @@
...
+            body=body, headers=self.build_headers(timestamp, body)
+        )
+        response = app.dispatch(request)
+        assert response.status == 404
+        assert self.mock_received_requests["/auth.test"] == 1
+
+    valid_reaction_added_body = {
+        "token": "verification_token",
+        "team_id": "T_SOURCE",
+        "enterprise_id": "E_SOURCE",
```
Comment (seratch) [inline tests/scenario_tests/test_events_shared_channels.py:140]: We used to us these values but they may not be the ones we need when an app works in shared channels.

File: tests/scenario_tests/test_events_shared_channels.py [hunk 6]
```diff
@@ -0,0 +1,514 @@
...
+            signing_secret=self.signing_secret,
+            authorize=authorize,
+        )
+
+        @app.event("app_mention")
+        def handle_app_mention(body, say: Say, payload, event):
+            assert body == self.valid_event_body
+            assert body["event"] == payload
+            assert payload == event
+            say("What's up?")
```
Comment (seratch) [inline tests/scenario_tests/test_events_shared_channels.py:100]: To make this functioning, I needed to improve `context.client`.

**Review by seratch — COMMENTED** [review]
Here are a few comments for reviewers

File: tests/scenario_tests/test_events_shared_channels.py [hunk 7]
```diff
@@ -0,0 +1,514 @@
...
+            "item_user": "W111",
+            "event_ts": "1599616881.000800",
+        },
+        "type": "event_callback",
+        "event_id": "Ev111",
+        "event_time": 1599616881,
+        "authorizations": [
+            {
+                "enterprise_id": "E_INSTALLED",
+                "team_id": "T_INSTALLED",
```
Comment (seratch) [inline tests/scenario_tests/test_events_shared_channels.py:156]: These fields should be used in any case. That's the main change in this PR.

File: tests/scenario_tests/test_events_shared_channels.py [hunk 8]
```diff
@@ -0,0 +1,514 @@
...
+    cleanup_mock_web_api_server,
+)
+from tests.utils import remove_os_env_temporarily, restore_os_env
+
+valid_token = "xoxb-valid"
+
+
+def authorize(enterprise_id, team_id, client: WebClient):
+    assert enterprise_id == "E_INSTALLED"
+    assert team_id == "T_INSTALLED"
```
Comment (seratch) [inline tests/scenario_tests/test_events_shared_channels.py:20]: These assertions fail with the current implementation.

File: tests/scenario_tests/test_events_shared_channels.py [hunk 9]
```diff
@@ -0,0 +1,514 @@
+import json
```
Comment (seratch) [inline tests/scenario_tests/test_events_shared_channels.py:1]: This test suite is copied from `test_events.py` and then is modified to emulate shared channel situations.

**Review by seratch — COMMENTED** [review]
added more comments for reviewers

**Review by stevengill — APPROVED** [review]

**Pull request merged**